python_modules/dagster/dagster/_core/definitions/metadata/__init__.py

```diff
@@ -1311,3 +1318,109 @@ class TableMetadataSet(NamespacedMetadataSet):
     @classmethod
     def namespace(cls) -> str:
         return "dagster"
+
+
+DEFAULT_SOURCE_FILE_KEY = "asset_definition"
+
+
+## Source Metadata
+
+
+@experimental
+@whitelist_for_serdes
+class LocalFileSource(DagsterModel):
+    """Represents a local file source location."""
+
+    file_path: str
+    line_number: int
+
+
+@experimental
+@whitelist_for_serdes
+class SourceMetadataValue(DagsterModel, MetadataValue["SourceMetadataValue"]):
+    """Metadata value type which represents source locations (locally or otherwise)
+    of the asset in question. For example, the file path and line number where the
+    asset is defined.
+
+    Properties:
+    sources (Dict[str, LocalFileSource]): A labeled dictionary of sources.
+    """
+
+    sources: Dict[str, LocalFileSource] = {}
+
+    @property
+    def value(self) -> "SourceMetadataValue":
+        return self
+
+
+def source_path_from_fn(fn: Callable[..., Any]) -> Optional[LocalFileSource]:
+    cwd = os.getcwd()
+    origin_file: Optional[str] = None
+    origin_line = None
+    try:
+        origin_file = os.path.abspath(os.path.join(cwd, inspect.getsourcefile(fn)))  # type: ignore
+        origin_file = check.not_none(origin_file)
+        origin_line = inspect.getsourcelines(fn)[1]
+    except TypeError:
+        return None
+
+    return LocalFileSource(
+        file_path=origin_file,
+        line_number=origin_line,
+    )
+
+
+class SourceDataMetadataSet(NamespacedMetadataSet):
+    """Metadata entries that apply to asset definitions and which specify the source code location
+    for the asset.
+    """
+
+    source_data: SourceMetadataValue
```

Comment:
When choosing a name here, we want to minimize the possibility of collisions with other metadata entries we might want to add in the future.  "Source" can refer a lot of things, e.g. the source data in a data pipeline.

Perhaps "source_code_links"? "code_links"?

I think that would imply different names for some of the other classes here as well.

Reply:
I opted for `SouceCodeLocationsMetadataValue` in this pass, let me know what you think.